lib/src/main/java/org/team100/lib/barcode/Sensor.java

```diff
@@ -0,0 +1,256 @@
+
+    /**
+     * @param input      all the sensors are multiplexed to one analog input.
+     * @param thresholds for each sensor, N of them.
+     * @param mux        the array uses N pins starting at zero.
+     * @param odd        turn on the odd emitters when high
+     * @param even       turn on the even emitters when high
+     */
+    public Sensor(
+            AnalogInput input,
+            double[] thresholds,
+            Mux mux,
+            DigitalOutput odd,
+            DigitalOutput even) {
+        m_input = input;
+        m_thresholds = thresholds;
+        m_mux = mux;
+        m_odd = odd;
+        m_even = even;
+        // the default state for the outputs should be off
+        m_odd.set(false);
+        m_even.set(false);
+    }
+
+    /** @return the number represented by the barcode. */
+    public OptionalInt readValue() {
+        boolean[] read = read();
+        if (read.length == 0)
+            return OptionalInt.empty();
+        return OptionalInt.of(toChannel(read));
+    }
+
+    /////////////////////////////////////////////////
+
+    /**
+     * @return array representing the state of the sensors, 1 means dark
+     */
+    boolean[] read() {
+        try {
+            boolean[] result = new boolean[m_thresholds.length];
+            // to read all the bits, we have to scan with the mux.
+            // first turn on the odd emitters, which actually means 0,2,4.
+            m_odd.set(true);
+            Thread.sleep(0, 85);// led driver startup time
+            Thread.sleep(0, 200);// sensor startup time minus the mux time below
+            for (int i = 0; i < m_thresholds.length; i += 2) {
+                m_mux.set(i);
+                Thread.sleep(5, 0);// 100ns of settling time
+                result[i] = readInput(m_thresholds[i], i);
+            }
+            // switch the emitters to 1,3 (note the confusing naming here)
+            m_odd.set(false);
+            m_even.set(true);
+            Thread.sleep(0, 85);// led driver startup time
+            Thread.sleep(0, 200);// sensor startup time minus the mux time below
+            for (int i = 1; i < m_thresholds.length; i += 2) {
+                m_mux.set(i);
+                Thread.sleep(5, 0);// 100ns of settling time
+                result[i] = readInput(m_thresholds[i], i);
+            }
+            m_even.set(false);
+            return result;
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+            return new boolean[0]; // empty means failure
+        } finally {
+            // make sure the emitters are off
+            m_odd.set(false);
+            m_even.set(false);
+        }
+    }
+
+    /**
+     * @param threshold above the threshold means dark, 1, below means light, is 0
+     * @return true = dark = "1" in the bar code.
+     */
+    boolean readInput(double threshold, int i) {
+        double sv = m_input.getAverageVoltage();
+        switch (i) {
+            case 0:
```

Comment:
could you explain a little about what you're doing here?  it looks like you're averaging over a sliding window with 150 samples?  could you use the WPILib filter for that?